eyelid-match-ops/src/primitives/poly.rs

```diff
@@ -136,3 +157,87 @@ pub fn poly_split(a: &Poly) -> (Poly, Poly) {
+/// Otherwise, returns an error.
+pub fn inverse(a: &Poly) -> Result<Poly, String> {
+    let mut mod_pol = Poly::zero();
+
+    mod_pol[MAX_POLY_DEGREE] = Coeff::one();
+    mod_pol[0] = Coeff::one();
+
+    let y = extended_gcd(&mod_pol, a);
+    let mul_both = a.clone().mul(y.clone());
+    if mul_both.is_one() {
+        Ok(y)
+    } else {
+        Err("No inverse in the ring.".to_owned())
+    }
+}
+
+/// Helps to calculate the equation `cur = prev - q.cur`.
+fn update_diophantine(prev: Poly, cur: Poly, q: Poly) -> (Poly, Poly) {
+    let aux = cur.clone();
+    let mul_res = cur.mul(&q.clone());
+    let new_cur = prev.sub(&mul_res).clone();
+    let new_prev = aux.clone();
+
+    (new_cur, new_prev)
+}
+
+/// Returns polynomials x, y, d such that a.x + b.y = d.
+/// When d=1 we have that x is the multiplicative inverse of a.
+pub fn extended_gcd(a: &Poly, b: &Poly) -> Poly {
+    // Invariant a.xi + b.yi = ri
+
+    // init with x0=1, y0=0, r0=a
+    let mut x_prev = Poly::one();
+    let mut y_prev = Poly::zero();
+    let mut ri_prev = a.clone();
+    // next:     x1=0, y1=1, r1=b
+    let mut x_cur = Poly::zero();
+    let mut y_cur = Poly::one();
+    let ri_cur = b.clone();
+
+    let ri_aux = ri_cur.clone();
+    // TODO: q is just a monomial, then we can optimize the next computation
+    let (mut q, mut ri_cur) = DenseOrSparsePolynomial::from(ri_prev.clone())
+        .divide_with_q_and_r(&ri_cur.into())
+        .expect("init divisor is not zero");
+    ri_prev = ri_aux;
+    // x_cur = x_prev - q.x_cur
+    (x_cur, x_prev) = update_diophantine(x_prev, x_cur, q.clone().into());
+    // y_cur = y_prev - q.y_cur
+    (y_cur, y_prev) = update_diophantine(y_prev, y_cur, q.clone().into());
+    // loop until ri_cur = 0
+    while !(ri_cur.is_zero()) {
+        let ri_aux = ri_cur.clone();
+        // TODO: q is just a monomial, then we can optimize the next computation
+        (q, ri_cur) = DenseOrSparsePolynomial::from(ri_prev.clone())
+            .divide_with_q_and_r(&ri_cur.into())
+            .expect("loop divisor is not zero");
+        ri_prev = ri_aux.into();
+        // x_cur = x_prev - q.x_cur
+        (x_cur, x_prev) = update_diophantine(x_prev, x_cur, q.clone().into());
+        // y_cur = y_prev - q.y_cur
+        (y_cur, y_prev) = update_diophantine(y_prev, y_cur, q.clone().into());
+    }
+    // compute ri_prev inverse to calculate the final result
+    let divisor = ri_prev.clone();
+    // FIXME: if b is not invertible mod a the assert is not going to pass.
+    // Since the test is probabilistic, we have a small chance of failure.
+    // It fails after a small number of executions.
+    // Sometimes it fails with a different error, meaning there is another
+    // problem happening.
+    // It occasionally fails when we do the following:
+    // export RUSTFLAGS="-D warnings --cfg tiny_poly"
+    // cargo test --all-targets -- --nocapture inv
+    debug_assert!(divisor.degree() == 0);
```

Comment:
If this assert is necessary for the method to return correct results, and it’s possible for the assert to fail on some random or user inputs, then we should return an error instead.

Assertions that prevent incorrect results can’t be `debug_assert!()`, because then we could get incorrect results in production. If an incorrect result is used in key generation, that is a security issue.

And they can’t be `assert!()` either, because panicking in production is a reliability bug.

(Assertions are only for things we think are always true. They help us find bugs in new or changed code, where what we thought was wrong.)

Reply:
Replaced by an error. Also removed the error treatment from inverse() function and the unnecessary check (now happening at the unit test). 

Since we have opened an [issue](https://github.com/Inversed-Tech/eyelid/issues/55) for better testing it, I think this solves the problem for now. Therefore I also removed the FIXME comment. Do you agree?